File: solutions/python/05/segtree_lazy/main.py
class LazySegTree:
    """区間加算・区間和（サイズは 2 の冪）"""

    def __init__(self, data: list) -> None:
        n = 1
        while n < len(data):
            n <<= 1
        self.n = n
        self.seg = [0] * (2 * n)
        self.lazy = [0] * (2 * n)
        for i, v in enumerate(data):
            self.seg[n + i] = v

    def _apply(self, i: int, l: int, r: int, x: int) -> None:
        self.seg[i] += x * (r - l)
        self.lazy[i] += x

    def _push(self, i: int, l: int, r: int) -> None:
        if self.lazy[i] == 0 or i >= self.n:
            return
        m = (l + r) // 2
        x = self.lazy[i]
        self._apply(2 * i, l, m, x)
        self._apply(2 * i + 1, m, r, x)
        self.lazy[i] = 0

    def _pull(self, i: int) -> None:
        self.seg[i] = self.seg[2 * i] + self.seg[2 * i + 1]

    def range_add(self, ql: int, qr: int, x: int, i: int = 1, l: int = 0, r: int | None = None) -> None:
        if r is None:
            r = self.n
        if qr <= l or r <= ql:
            return
        if ql <= l and r <= qr:
            self._apply(i, l, r, x)
            return
        self._push(i, l, r)
        m = (l + r) // 2
        self.range_add(ql, qr, x, 2 * i, l, m)
        self.range_add(ql, qr, x, 2 * i + 1, m, r)
        self._push(2 * i, l, m)
        self._push(2 * i + 1, m, r)
        self._pull(i)

    def range_sum(self, ql: int, qr: int, i: int = 1, l: int = 0, r: int | None = None) -> int:
        if r is None:
            r = self.n
        if qr <= l or r <= ql:
            return 0
        if ql <= l and r <= qr:
            return self.seg[i]
        self._push(i, l, r)
        m = (l + r) // 2
        return self.range_sum(ql, qr, 2 * i, l, m) + self.range_sum(ql, qr, 2 * i + 1, m, r)


def main() -> None:
    st = LazySegTree([0] * 8)
    st.range_add(1, 5, 10)
    st.range_add(3, 7, 5)
    print(st.range_sum(0, 8))


if __name__ == "__main__":
    main()

File: solutions/go/05/segtree_lazy/main.go
package main

import "fmt"

type LazySeg struct {
	n, size int
	seg     []int64
	lazy    []int64
}

func NewLazySeg(data []int) *LazySeg {
	size := 1
	for size < len(data) {
		size <<= 1
	}
	seg := make([]int64, 2*size)
	lazy := make([]int64, 2*size)
	for i, v := range data {
		seg[size+i] = int64(v)
	}
	return &LazySeg{len(data), size, seg, lazy}
}

func (t *LazySeg) apply(i, l, r int, x int64) {
	t.seg[i] += x * int64(r-l)
	t.lazy[i] += x
}

func (t *LazySeg) push(i, l, r int) {
	if t.lazy[i] == 0 || i >= t.size {
		return
	}
	m := (l + r) / 2
	x := t.lazy[i]
	t.apply(2*i, l, m, x)
	t.apply(2*i+1, m, r, x)
	t.lazy[i] = 0
}

func (t *LazySeg) pull(i int) {
	t.seg[i] = t.seg[2*i] + t.seg[2*i+1]
}

func (t *LazySeg) RangeAdd(ql, qr int, x int64) {
	t.rangeAdd(ql, qr, x, 1, 0, t.size)
}

func (t *LazySeg) rangeAdd(ql, qr int, x int64, i, l, r int) {
	if qr <= l || r <= ql {
		return
	}
	if ql <= l && r <= qr {
		t.apply(i, l, r, x)
		return
	}
	t.push(i, l, r)
	m := (l + r) / 2
	t.rangeAdd(ql, qr, x, 2*i, l, m)
	t.rangeAdd(ql, qr, x, 2*i+1, m, r)
	t.push(2*i, l, m)
	t.push(2*i+1, m, r)
	t.pull(i)
}

func (t *LazySeg) RangeSum(ql, qr int) int64 {
	return t.rangeSum(ql, qr, 1, 0, t.size)
}

func (t *LazySeg) rangeSum(ql, qr, i, l, r int) int64 {
	if qr <= l || r <= ql {
		return 0
	}
	if ql <= l && r <= qr {
		return t.seg[i]
	}
	t.push(i, l, r)
	m := (l + r) / 2
	return t.rangeSum(ql, qr, 2*i, l, m) + t.rangeSum(ql, qr, 2*i+1, m, r)
}

func main() {
	st := NewLazySeg(make([]int, 8))
	st.RangeAdd(1, 5, 10)
	st.RangeAdd(3, 7, 5)
	fmt.Println(st.RangeSum(0, 8))
}
